pkg/controllers/cluster/cluster_controller_test.go

```diff
@@ -0,0 +1,250 @@
+			cluster: &clusterv1alpha1.Cluster{
+				ObjectMeta: controllerruntime.ObjectMeta{
+					Name:       "test-cluster",
+					Finalizers: []string{util.ClusterControllerFinalizer},
+				},
+			},
+			want:    controllerruntime.Result{},
+			wantErr: false,
+		},
+		{
+			name: "cluster with ready condition",
+			cluster: &clusterv1alpha1.Cluster{
+				ObjectMeta: controllerruntime.ObjectMeta{
+					Name:       "test-cluster",
+					Finalizers: []string{util.ClusterControllerFinalizer},
+				},
+				Status: clusterv1alpha1.ClusterStatus{
+					Conditions: []metav1.Condition{
+						{
+							Type:   clusterv1alpha1.ClusterConditionReady,
+							Status: metav1.ConditionTrue,
+						},
+					},
+				},
+			},
+			want:    controllerruntime.Result{},
+			wantErr: false,
+		},
+		{
+			name: "cluster not found",
+			cluster: &clusterv1alpha1.Cluster{
+				ObjectMeta: controllerruntime.ObjectMeta{
+					Name:       "test-cluster-noexist",
+					Finalizers: []string{util.ClusterControllerFinalizer},
+				},
+			},
+			want:    controllerruntime.Result{},
+			wantErr: false,
+		},
+		{
+			name: "remove cluster failed",
+			cluster: &clusterv1alpha1.Cluster{
+				ObjectMeta: controllerruntime.ObjectMeta{
+					Name:       "test-cluster",
+					Finalizers: []string{util.ClusterControllerFinalizer},
+				},
+				Spec: clusterv1alpha1.ClusterSpec{
+					SyncMode: clusterv1alpha1.Pull,
+				},
+				Status: clusterv1alpha1.ClusterStatus{
+					Conditions: []metav1.Condition{
+						{
+							Type:   clusterv1alpha1.ClusterConditionReady,
+							Status: metav1.ConditionFalse,
+						},
+					},
+				},
+			},
+			ns: &corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: names.GenerateExecutionSpaceName("test-cluster"),
+				},
+			},
+			work: &workv1alpha1.Work{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:       "test-work",
+					Namespace:  names.GenerateExecutionSpaceName("test-cluster"),
+					Finalizers: []string{util.ExecutionControllerFinalizer},
+				},
+			},
+			del:     true,
+			want:    controllerruntime.Result{},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c := newClusterController()
+			if tt.cluster != nil {
+				if err := c.Create(context.Background(), tt.cluster, &client.CreateOptions{}); err != nil {
+					t.Fatalf("faild to create cluster %v", err)
+				}
+			}
+
+			if tt.ns != nil {
+				if err := c.Create(context.Background(), tt.ns, &client.CreateOptions{}); err != nil {
+					t.Fatalf("faild to create ns %v", err)
+				}
+			}
+
+			if tt.work != nil {
+				if err := c.Create(context.Background(), tt.work, &client.CreateOptions{}); err != nil {
+					t.Fatalf("faild to create work %v", err)
+				}
+			}
+
+			if tt.del {
+				if err := c.Delete(context.Background(), tt.cluster, &client.DeleteOptions{}); err != nil {
+					t.Fatalf("failed to delete cluster %v", err)
+				}
+			}
+
+			got, err := c.Reconcile(context.Background(), req)
```

Comment:
In addition to checking err and result, can you determine whether cluster resources are expected?

Other tests are similar.